reactive/kotlinx-coroutines-rx3/src/RxAwait.kt

```diff
@@ -42,7 +42,7 @@ public suspend fun CompletableSource.await(): Unit = suspendCancellableCoroutine
  * function immediately resumes with [CancellationException] and disposes of its subscription.
  */
 @Suppress("UNCHECKED_CAST")
-public suspend fun <T> MaybeSource<T>.awaitSingleOrNull(): T? = suspendCancellableCoroutine { cont ->
+public suspend fun <T : Any> MaybeSource<T>.awaitSingleOrNull(): T? = suspendCancellableCoroutine { cont ->
```

Comment:
tl;dr: we'll probably just make this compile somehow, and only make the proposed change *after* 1.7.

Adding `T : Any` boundaries is a good change. In fact, it's so good that it's the third time it's suggested! The reason our earlier attempts failed is that this would break a lot of code generic in `T` that depends on this library: for example, https://github.com/Kotlin/kotlinx.coroutines/pull/2630. It looks like the change to the compiler will force our hand, but the general concern still stands: we don't want to introduce breaking changes that aren't that useful.

Due to this, I think the proper thing to do here is to make the code compile *without* introducing any incompatibility. When people migrate their code to 1.7, they will in any case have to adapt their code a lot, including the generic functions that we don't want to break. After they fix this on their side, we'll be able to remove the hacks that we'll need for the code to compile, and without breaking any code.

What do you think? I didn't understand from your description whether you actually need this change already for some reason.

Reply:
By the way, there is a way to workaround that: https://github.com/Kotlin/kotlinx.coroutines/pull/2954 
It's not currently merged due to IDE issues

Reply:
Essentially, my _goal_ is to break a bunch of Google's code that depends on this library :)

Of course, I don't really _want_ to break our code. But as you say, I expect for us to _have_ to break it eventually. So I would rather break (well, fix) a small amount of code now than I would break (fix) a large amount of code later.

To be fair, this is a particularly good time for a breaking change _for us_: We don't have a ton of RxJava3 code _yet_, but we're expecting to start seeing a lot more soon. So it was pretty easy to update our existing users of these APIs a couple weeks ago, but already I'm seeing backsliding.

I understand that, in the wider community, these APIs already have far more users. Still, I think the principle stands: Updating users of the APIs is only going to get harder over time.

It would be great if we could avoid the breaking change entirely with #2954. However, I suspect that that's not going to work in the long term: Even an _API declaration_ like [`public suspend fun <T> SingleSource<T>.await()`](https://github.com/Kotlin/kotlinx.coroutines/pull/2954/files#diff-7c46154c0f1691c6a9d7804e3e19769f0f383b1894ee96b9e716889e8c2f68c5R117) is going to become an error eventually, since [`SingleSource` requires a non-null type argument](https://github.com/ReactiveX/RxJava/blob/e1e35b05030ee0b8bc5e7b1b5c7e47b8d7617019/src/main/java/io/reactivex/rxjava3/core/SingleSource.java#L29). So I think that even that API has to change to `public suspend fun <T> SingleSource<T`**` & Any`**`>.await()`. And I think that's as much a breaking change as this PR is.

---

I _may_ be able to mostly accomplish my goal another way: If I can set the `@io.reactivex.rxjava3.annotations:strict` flag for the rest of our codebase, then I _may_ be able to prevent callers from ever using these APIs with a nullable type argument. As it turns out, though, that's tricky because, if I set that flag for our whole codebase, we can no longer compile this project :) And our build setup / kotlinc doesn't appear to provide a good way to turn a flag "off" if it's on by default, so I can't easily turn it off just for this project. I can continue exploring ways to make that work, too; I'm just trying as many different approaches as I can.

Reply:
Two small updates on the breaking-changes front:
- First, if you ultimately find this PR to be too disruptive, it looks like we'll probably maintain it as a small patch for Google's copy. That unblocks my goal of flipping [-Xnullability-annotations=@io.reactivex.rxjava3.annotations:strict](https://kotlinlang.org/docs/java-interop.html#nullability-annotations) on for our depot. (I very much appreciate your catching my earlier errors so that I didn't break something, even if it would only have been in Google's code and not for _all_ rxjava3 users :))
- We've found that another future Kotlin flag flip, -Xenhance-type-parameter-types-to-def-not-null=true, forces our rxjava3 users to make some (but not all) of the same changes as this PR does. I don't know if that flag and the rxjava3.annotations flag will end up getting flipped by default in the same release of Kotlin or different releases. Nor do I know whether users would prefer to take all their rxjava3-related breakages at once or have multiple smaller breaking releases. But I wanted to pass along that that's another Kotlin future flag flip that may break some kotlinx-coroutines users.